src/main/scala/sigmastate/utxo/transformers.scala

```diff
@@ -270,67 +299,88 @@ case class ExtractScriptBytes(input: Value[SBox.type]) extends Extract[SByteArra
 
-  override lazy val cost: Int = 1000 //todo: make it PerKb * max box size in kbs
+  override def cost[C <: Context[C]](context: C): Long = 1000 //todo: make it PerKb * max box size in kbs
 
   override def function(box: EvaluatedValue[SBox.type]): Value[SByteArray.type] = ByteArrayConstant(box.value.bytes)
 }
 
 case class ExtractBytesWithNoRef(input: Value[SBox.type]) extends Extract[SByteArray.type] with NotReadyValueByteArray {
   override val opCode: OpCode = OpCodes.ExtractBytesWithNoRefCode
 
-  override lazy val cost: Int = 1000 //todo: make it PerKb * max box size in kbs
+  override def cost[C <: Context[C]](context: C) = 1000 //todo: make it PerKb * max box size in kbs
 
   override def function(box: EvaluatedValue[SBox.type]): Value[SByteArray.type] = ByteArrayConstant(box.value.bytesWithNoRef)
 }
 
 case class ExtractId(input: Value[SBox.type]) extends Extract[SByteArray.type] with NotReadyValueByteArray {
   override val opCode: OpCode = OpCodes.ExtractIdCode
 
-  override lazy val cost: Int = 10
+  override def cost[C <: Context[C]](context: C) = 10
 
   override def function(box: EvaluatedValue[SBox.type]): Value[SByteArray.type] = ByteArrayConstant(box.value.id)
 }
 
 case class ExtractRegisterAs[V <: SType](input: Value[SBox.type],
                                          registerId: RegisterIdentifier,
                                          default: Option[Value[V]] = None)(implicit val tpe: V)
-    extends Extract[V] with NotReadyValue[V] with Rewritable {
+  extends Extract[V] with NotReadyValue[V] with Rewritable {
   override val opCode: OpCode = OpCodes.ExtractRegisterAs
 
   def arity = 4
+
   def deconstruct = immutable.Seq[Any](input, registerId, default, tpe)
+
   def reconstruct(cs: immutable.Seq[Any]) = cs match {
-    case Seq(input: Value[SBox.type] @unchecked,
-      registerId: RegisterIdentifier,
-      default: Option[Value[V]] @unchecked,
-      t: V @unchecked) => ExtractRegisterAs[V](input, registerId, default)(t)
+    case Seq(input: Value[SBox.type]@unchecked,
+    registerId: RegisterIdentifier,
+    default: Option[Value[V]]@unchecked,
+    t: V@unchecked) => ExtractRegisterAs[V](input, registerId, default)(t)
     case _ =>
       illegalArgs("ExtractRegisterAs", "(Value[SBox.type], registerId: RegisterIdentifier, default: Option[Value[V]])(tpe: V)", cs)
   }
-  override def cost: Int = 1000 //todo: the same as ExtractBytes.cost
+
+  override def cost[C <: Context[C]](context: C) = 1000 //todo: the same as ExtractBytes.cost
 
   override def function(box: EvaluatedValue[SBox.type]): Value[V] =
     box.value.get(registerId).orElse(default).get.asInstanceOf[Value[V]]
 }
 
+trait Deserialize[V <: SType] extends NotReadyValue[V] with Rewritable
 
 
-case class Deserialize[V <: SType](input: Value[SByteArray.type])(implicit val tpe: V)
-  extends Transformer[SByteArray.type, V] with NotReadyValue[V] with Rewritable {
+case class DeserializeContext[V <: SType](id: Byte)(implicit val tpe: V)
```

Comment:
@kushti implicit parameters in case classes require explicit implementation of `equals` and `hashCode`